compiler/passes/src/dead_code_elimination/dead_code_eliminator.rs

```diff
@@ -14,25 +14,89 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use leo_ast::NodeBuilder;
-use leo_span::Symbol;
+use leo_ast::{AccessExpression, Expression, Location, Node, Type};
+use leo_span::{Symbol, sym};
 
 use indexmap::IndexSet;
 
+use crate::{SymbolTable, TypeTable};
+
+#[derive(Debug)]
 pub struct DeadCodeEliminator<'a> {
-    /// A counter to generate unique node IDs.
-    pub(crate) node_builder: &'a NodeBuilder,
     /// The set of used variables in the current function body.
     pub(crate) used_variables: IndexSet<Symbol>,
-    /// Whether or not the variables are necessary.
-    pub(crate) is_necessary: bool,
-    /// Whether or not we are currently traversing an async function.
-    pub(crate) is_async: bool,
+
+    /// The name of the program currently being processed.
+    pub(crate) program_name: Symbol,
+
+    pub(crate) symbol_table: &'a SymbolTable,
+
+    pub(crate) type_table: &'a TypeTable,
 }
 
 impl<'a> DeadCodeEliminator<'a> {
     /// Initializes a new `DeadCodeEliminator`.
-    pub fn new(node_builder: &'a NodeBuilder) -> Self {
-        Self { node_builder, used_variables: Default::default(), is_necessary: false, is_async: false }
+    pub(crate) fn new(symbol_table: &'a SymbolTable, type_table: &'a TypeTable) -> Self {
+        Self { used_variables: Default::default(), program_name: Symbol::intern(""), symbol_table, type_table }
+    }
+
+    fn contains_future_or_record(&self, ty: &Type) -> bool {
+        use Type::*;
+        match ty {
+            Array(array) => self.contains_future_or_record(array.element_type()),
```

Comment:
In snarkVM, arrays can only contain other `Plaintext` types. Does TYC enforce this as well?